### PR TITLE
S3 to S3 Replication in SAM

### PR DESCRIPTION
**Pull request by jorgetovar**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by bfreiberg** [comment]
Hi @jorgetovar , thanks for the pull request. However, your submission does not represent a serverless pattern as we see it. Those are typically intended to be primarily IaC-focused implementations of 2-4 AWS services. Feel free to expand upon this and resubmit then.

**Comment by jorgetovar** [comment]
But I see a similar pattern here but with CDK

https://serverlessland.com/patterns/s3-s3-cdk

**Comment by bfreiberg** [comment]
Thanks for your feedback. We've discussed it and will accept this pattern, also given the pattern you referenced. I'll review your submission soon.

File: s3-s3-replication-sam/README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# AWS S3 Replication Pattern
+
+This pattern enables you to replicate objects across Amazon S3 buckets, providing a reliable and scalable solution for data redundancy and disaster recovery.
+
+**Key Benefits**:
```
Comment (bfreiberg) [inline s3-s3-replication-sam/README.md:5]: This section is a bit verbose and talks more about the feature than the pattern itself

File: s3-s3-replication-sam/README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    * Enter the desired AWS Region.
+    * Allow SAM CLI to create IAM roles with the required permissions.
+
+    Once you have run `sam deploy --guided` mode once and saved arguments to a configuration file (samconfig.toml), you can use `sam deploy` in the future to use these defaults.
+
+1. Note the outputs from the SAM deployment process. These contain the resource names and/or ARNs used for testing.
+
+## How it Works
+
+Explain how the service interaction works.
```
Comment (bfreiberg) [inline s3-s3-replication-sam/README.md:55]: This seems to be missing

File: s3-s3-replication-sam/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+## Testing
+
+Provide steps to trigger the integration and show what should be observed if successful.
```
Comment (bfreiberg) [inline s3-s3-replication-sam/README.md:59]: Please provide testing instructions

File: s3-s3-replication-sam/README.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+## Cleanup
+ 
+1. Delete the stack:
+    ```bash
+    aws cloudformation delete-stack --stack-name STACK_NAME
```
Comment (bfreiberg) [inline s3-s3-replication-sam/README.md:65]: You've used `sam` before. Why is now CloudFormation called directly?

File: s3-s3-replication-sam/README.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    ```
+1. Confirm the stack has been deleted:
+    ```bash
+    aws cloudformation list-stacks --query "StackSummaries[?contains(StackName,'STACK_NAME')].StackStatus"
+    ```
+
+----
+
+Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (bfreiberg) [inline s3-s3-replication-sam/README.md:74]:
```suggestion
Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: s3-s3-replication-sam/example-pattern.json [hunk 6]
```diff
@@ -0,0 +1,67 @@
+{
+  "title": "S3 to S3 replication",
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:2]:
```suggestion
  "title": "AWS S3 bucket replication",
```

File: s3-s3-replication-sam/example-pattern.json [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  "description": "Create a stack template that you can use to replicate objects between S3 buckets across AWS accounts and regions.",
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:3]:
```suggestion
  "description": "This pattern sets up replication between S3 buckets.",
```

File: s3-s3-replication-sam/example-pattern.json [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  "language": "Python",
+  "level": "200",
+  "framework": "SAM",
+  "introBox": {
+    "headline": "How it works",
+    "text": [
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:9]: Please keep this shorter, especially because the pattern does not set up cross-region or cross-account replication

File: s3-s3-replication-sam/example-pattern.json [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+      "Flexible Implementation: Leverage native S3 replication features to configure replication behavior through JSON-based, structured language."
+    ]
+    
+  },
+  "gitHub": {
+    "template": {
+      "repoURL": "https://github.com/aws-samples/serverless-patterns/tree/main/s3-s3-replication-sam",
+      "templateURL": "serverless-patterns/s3-s3-replication-sam",
+      "projectFolder": "s3-s3-replication-sam",
+      "templateFile": "s3-s3-replication-sam.py"
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:28]: That file is not part of this PR, did you forget to check it in?

File: s3-s3-replication-sam/example-pattern.json [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    }
+  },
+  "resources": {
+    "bullets": [
+      {
+        "text": "Call Athena with Step Functions",
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:34]: How is this relevant?

File: s3-s3-replication-sam/example-pattern.json [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        "link": "https://docs.aws.amazon.com/step-functions/latest/dg/connect-athena.html"
+      },
+      {
+        "text": "Amazon Athena - Serverless Interactive Query Service",
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:38]: How is this relevant?

File: s3-s3-replication-sam/example-pattern.json [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    ]
+  },
+  "testing": {
+    "text": [
+      "See the GitHub repo for detailed testing instructions."
+    ]
+  },
+  "cleanup": {
+    "text": [
+      "Delete the stack: <code>sam delete</code>."
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:55]:
```suggestion
      "<code>sam delete</code>"
```

File: s3-s3-replication-sam/template.yaml [hunk 13]
```diff
@@ -0,0 +1,97 @@
+AWSTemplateFormatVersion: '2010-09-09'
+Transform: 'AWS::Serverless-2016-10-31'
+Description: 'S3 Replication Template'
```
Comment (bfreiberg) [inline s3-s3-replication-sam/template.yaml:3]:
```suggestion
Description: 'S3 bucket replication'
```

File: s3-s3-replication-sam/template.yaml [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+                Action:
+                  - 's3:GetBucketVersioning'
+                  - 's3:ListBucket'
+                  - s3:GetReplicationConfiguration
+                  - s3:GetObjectVersionForReplication
+                  - s3:GetObjectVersionAcl
+                  - s3:GetObjectVersionTagging
+                  - s3:GetObjectRetention
+                  - s3:GetObjectLegalHold
+                Resource: '*'
```
Comment (bfreiberg) [inline s3-s3-replication-sam/template.yaml:42]: Please apply least-privilege and restrict this policy to the relevant buckets

File: s3-s3-replication-sam/template.yaml [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+              - Effect: 'Allow'
+                Action:
+                  - 's3:ReplicateObject'
+                  - 's3:ReplicateDelete'
+                  - 's3:ReplicateTags'
+                  - 's3:GetObjectVersionTagging'
+                  - 's3:ObjectOwnerOverrideToBucketOwner'
+                
+                Resource: '*'
```
Comment (bfreiberg) [inline s3-s3-replication-sam/template.yaml:51]: same as above

**Comment by bfreiberg** [comment]
@jorgetovar , do you still plan to work on this issue?

File: s3-s3-replication-sam/example-pattern.json (hunk 8 above)
Comment (jorgetovar) [reply s3-s3-replication-sam/example-pattern.json:9]: yes, thanks

**Review by jorgetovar — COMMENTED** [review]
Thanks, @bfreiberg, for all the suggestions. They have been addressed in the pull request now.

**Comment by jorgetovar** [comment]
@bfreiberg 

Please let me know if there are any specific areas you'd like me to address or if you need any further information from my end.

Thank you for your time and good advice

File: s3-s3-replication-sam/example-pattern.json (hunk 8 above)
Comment (bfreiberg) [reply s3-s3-replication-sam/example-pattern.json:9]: Hi, this section is still very verbose. Please keep it a bit shorter and focus on _"how it works"_. Something like: _"This pattern sets up S3 replication between..."_

File: s3-s3-replication-sam/example-pattern.json [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+  },
+  "resources": {
+    "bullets": [
+      {
+        "text": "Replicating Objects",
+        "link": "https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication.html"
+      },
+      {
+        "text": "Cross Region Replication",
+        "link": "https://www.youtube.com/watch?v=_6admeBN-lI"
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:32]: This pattern only sets up replication within the same region. Please include a more relevant link

File: s3-s3-replication-sam/example-pattern.json [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+    "text": [
+      "<code>sam delete</code>"
+    ]
+  },
+  "authors": [
+    {
+      "name": "Jorge Tovar",
+      "image": "https://avatars.githubusercontent.com/u/21094604?v=4",
+      "bio": "Software Architect, passionate about serverless technologies and cloud computing.",
+      "linkedin": "www.linkedin.com/in/jorgetovar-sa",
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:56]:
```suggestion
      "linkedin": "jorgetovar-sa",
```

**Comment by bfreiberg** [comment]
Looks good in general. I've highlighted a few minor adjustments

File: s3-s3-replication-sam/example-pattern.json (hunk 8 above)
Comment (jorgetovar) [reply s3-s3-replication-sam/example-pattern.json:9]: Ok updated :)

File: s3-s3-replication-sam/example-pattern.json (hunk 16 above)
Comment (jorgetovar) [reply s3-s3-replication-sam/example-pattern.json:32]: yes, you right. solved

**Comment by jorgetovar** [comment]
@bfreiberg

Please let me know if there are anything else to address or if you need any further information from my end.

**Comment by jorgetovar** [comment]
> Looks good in general. I've highlighted a few minor adjustments 

Looking forward to see if it's anything else missing

File: s3-s3-replication-sam/example-pattern.json [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+  },
+  "resources": {
+    "bullets": [
+      {
+        "text": "Replicating Objects",
+        "link": "https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication.html"
+      },
+      {
+        "text": "Cross Region Replication",
+        "link": "https://aws.amazon.com/about-aws/whats-new/2019/09/amazon-s3-introduces-same-region-replication/"
```
Comment (bfreiberg) [inline s3-s3-replication-sam/example-pattern.json:29]: I would prefer a link to the documentation (https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication.html) instead of a 5 year old announcement. The announcement is still valid but I believe the documentation provides much more context directly. What do you think?

**Comment by bfreiberg** [comment]
Looks pretty good, thanks for the updates. I highlighted two minor changes. It would be great if you could address them

File: s3-s3-replication-sam/example-pattern.json (hunk 18 above)
Comment (jorgetovar) [reply s3-s3-replication-sam/example-pattern.json:29]: I have updated this with a Youtube video link, the one that you send is already in the json

**Comment by jorgetovar** [comment]
I just had my son, so I was a bit delayed. Link updated. Everything should be ready for the merge now. Thank you for the assistance

@bfreiberg

**Comment by jorgetovar** [comment]
Hello looking forward for any updates:)

**Comment by bfreiberg** [comment]
I'm really sorry but I missed the incorrect service name before. Once that is changed, the PR is good to go

**Comment by bfreiberg** [comment]
Hi @jorgetovar , the service name still hasn't been changed. Are you sure that you pushed your latest changes?

**Comment by bfreiberg** [comment]
Great, thanks @jorgetovar for your contribution. A developer advocate will merge your PR soon

**Review by bfreiberg — APPROVED** [review]
